### PR TITLE
[libelf] Add new port

### PR DESCRIPTION
**Pull request by AlexAltea**
This partly takes care of issue #3790, even though its author inexplicably closed the issue. :-)

### CODE REVIEW

**Comment by Voskrese** [comment]
I made a port from here and closed it)
https://github.com/avast/libdwarf

File: ports/libelf/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        ${SOURCE_PATH}/lib/config.h.w32
+        ${SOURCE_PATH}/lib/config.h
+        COPYONLY)
+    configure_file(
+        ${SOURCE_PATH}/lib/sys_elf.h.w32
+        ${SOURCE_PATH}/lib/sys_elf.h
+        COPYONLY)
+endif()
+
+set(VCPKG_POLICY_DLLS_WITHOUT_LIBS enabled)
```
Comment (ras0219-msft) [inline ports/libelf/portfile.cmake]: This policy is almost never applicable. In this case, I would guess that the library does not support building a DLL for windows.

Assuming that is true, this should be removed and this file should call `vcpkg_check_linkage(ONLY_STATIC_LIBRARY)` before downloading the sources.

File: ports/libelf/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,41 @@
+cmake_minimum_required(VERSION 3.0)
+project(libelf C)
+
+file(GLOB_RECURSE SOURCES "${PROJECT_SOURCE_DIR}/*.c")
+add_library(libelf SHARED ${SOURCES})
```
Comment (ras0219-msft) [inline ports/libelf/CMakeLists.txt]: See below -- the `SHARED` keyword should probably be removed.

File: ports/libelf/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,46 @@
+include(vcpkg_common_functions)
+
+set(LIB_VERSION 0.8.13)
+set(LIB_FILENAME libelf_${LIB_VERSION}.orig.tar.gz)
+
+vcpkg_download_distfile(ARCHIVE
+    URLS "http://deb.debian.org/debian/pool/main/libe/libelf/${LIB_FILENAME}"
```
Comment (ras0219-msft) [inline ports/libelf/portfile.cmake]: Let's use `https://web.archive.org/web/20181111033959/www.mr511.de/software/libelf-0.8.13.tar.gz` as homebrew[1] does.

[1] https://github.com/Homebrew/homebrew-core/blob/master/Formula/libelf.rb

File: ports/libelf/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,41 @@
+cmake_minimum_required(VERSION 3.0)
+project(libelf C)
+
+file(GLOB_RECURSE SOURCES "${PROJECT_SOURCE_DIR}/*.c")
+add_library(libelf SHARED ${SOURCES})
+target_include_directories(libelf PRIVATE ${PROJECT_SOURCE_DIR})
+
+if(MSVC)
+    target_compile_definitions(libelf PUBLIC HAVE_CONFIG_H)
+    set_target_properties(libelf PROPERTIES WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
```
Comment (ras0219-msft) [inline ports/libelf/CMakeLists.txt]: As a policy[1], we do not use `WINDOWS_EXPORT_ALL_SYMBOLS` to create shared libraries on Windows.

If it is critical for this library to be shared, we should instead add a `.def` file with explicit symbol exports.

[1] https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#do-not-add-cmake_windows_export_all_symbols

File: ports/libelf/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,41 @@
+cmake_minimum_required(VERSION 3.0)
+project(libelf C)
+
+file(GLOB_RECURSE SOURCES "${PROJECT_SOURCE_DIR}/*.c")
+add_library(libelf SHARED ${SOURCES})
+target_include_directories(libelf PRIVATE ${PROJECT_SOURCE_DIR})
```
Comment (ras0219-msft) [inline ports/libelf/CMakeLists.txt]: Should probably also have `PUBLIC $<INSTALL_INTERFACE:include>` to ensure downstream users can find the headers when using `find_package()`.

File: ports/libelf/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+# Generate and install libelfConfig.cmake
+configure_package_config_file("Config.cmake.in" "${PROJECT_CONFIG}"
+    INSTALL_DESTINATION "${CONFIG_INSTALL_DIR}")
+install(FILES "${PROJECT_CONFIG}"
+    DESTINATION "${CONFIG_INSTALL_DIR}")
+
+# Generate and install libelfTargets*.cmake
+install(EXPORT ${TARGETS_EXPORT_NAME}
+    DESTINATION "${CONFIG_INSTALL_DIR}")
```
Comment (ras0219-msft) [inline ports/libelf/CMakeLists.txt]: `install(EXPORT)` should almost always be passed a `NAMESPACE` argument. In this case, `NAMESPACE libelf::`.

**Review by ras0219-msft — CHANGES_REQUESTED** [review]

**Comment by ras0219-msft** [comment]
Closing for inactivity; please open a new PR if you'd like to continue!

**Pull request closed without merging**